options-manager.js

```diff
@@ -0,0 +1,179 @@
+'use strict';
+var path = require('path');
+var arrify = require('arrify');
+var pkgConf = require('pkg-conf');
+var deepAssign = require('deep-assign');
+var resolveFrom = require('resolve-from');
+var objectAssign = require('object-assign');
+var homeOrTmp = require('home-or-tmp');
+var mutlimatch = require('multimatch');
+
+var DEFAULT_IGNORE = [
+	'node_modules/**',
+	'bower_components/**',
+	'coverage/**',
+	'{tmp,temp}/**',
+	'**/*.min.js',
+	'**/bundle.js',
+	'fixture{-*,}.{js,jsx}',
+	'{test/,}fixture{s,}/**',
+	'vendor/**',
+	'dist/**'
+];
+
+var DEFAULT_CONFIG = {
+	useEslintrc: false,
+	cache: true,
+	cacheLocation: path.join(homeOrTmp, '.xo-cache/'),
+	baseConfig: {
+		extends: 'xo'
+	}
+};
+
+var DEFAULT_PLUGINS = [
+	'no-empty-blocks'
+	// https://github.com/dustinspecker/eslint-plugin-no-use-extend-native/issues/16
+	// 'no-use-extend-native'
+];
+
+function normalizeOpts(opts) {
+	opts = objectAssign({}, opts);
+	// alias to help humans
+	['env', 'global', 'ignore', 'plugin', 'rule', 'extend'].forEach(function (singular) {
+		var plural = singular + 's';
+		var value = opts[plural] || opts[singular];
+
+		delete opts[singular];
+
+		if (value === undefined) {
+			return;
+		}
+
+		if (singular !== 'rule') {
+			value = arrify(value);
+		}
+
+		opts[plural] = value;
+	});
+
+	return opts;
+}
+
+function mergeWithPkgConf(opts) {
+	opts = objectAssign({
+		cwd: process.cwd()
+	}, opts);
+
+	return objectAssign({}, pkgConf.sync('xo', opts.cwd), opts);
+}
+
+function buildConfig(opts) {
+	var config = deepAssign({}, DEFAULT_CONFIG, {
+		envs: opts.envs,
+		globals: opts.globals,
+		plugins: DEFAULT_PLUGINS.concat(opts.plugins || []),
+		rules: opts.rules,
+		fix: opts.fix
+	});
+
+	if (!config.rules) {
+		config.rules = {};
+	}
+
+	if (opts.space) {
+		var spaces = typeof opts.space === 'number' ? opts.space : 2;
+		config.rules.indent = [2, spaces, {SwitchCase: 1}];
+	}
+
+	if (opts.semicolon === false) {
+		config.rules.semi = [2, 'never'];
+		config.rules['semi-spacing'] = [2, {before: false, after: true}];
+	}
+
+	if (opts.esnext) {
+		config.baseConfig.extends = 'xo/esnext';
+	} else {
+		// always use the Babel parser so it won't throw
+		// on esnext features in normal mode
+		config.parser = 'babel-eslint';
+		config.plugins = ['babel'];
```

Comment:
Also, I noticed that if `opts.esnext` is not `true`, we overwrite any plugins they might have specified. Should we just be appending/prepending babel to the list?